### DISCUSSION TITLE
General inquiries

### DISCUSSION THREAD

**Discussion by barbosa89** [zephir-lang/zephir#2314] [Q&A]
Hi,

I have known Zephir for a long time, and I would like to ask you a few questions about things that let me sleep:

- Can you use PHP core classes? I understand that functions can be used, are there any restrictions?
- Is it possible to rebuild things like Swoole completely with Zephir?
- Is it possible to use binaries like FFMPEG?
- Is it possible to use Composer packages?
- Is it possible to import other extensions?

I would love to know about the possibilities to start dreaming and creating things. Thanks.

**Comment by Jeckerson** [comment]
Hello

> Can you use PHP core classes? I understand that functions can be used, are there any restrictions?

Yes, if it is internal or if it is from core extension, full list can be found here:
https://github.com/zephir-lang/zephir/blob/d5bcff3476ef2004ef7aaadc561282c07d068a2d/config/class-entries.php#L34

> Is it possible to rebuild things like Swoole completely with Zephir?

Not completely, as Zephir doesn't support some features, especially async (work in progress). But extension made by Zephir can be easily used with Swoole inside PHP project

> Is it possible to use binaries like FFMPEG?

Any external C libraries can be included inside your project via Custom optimizers (https://docs.zephir-lang.com/0.12/en/optimizers)

> Is it possible to use Composer packages?

No, because your code is compiled and shipped as PHP extension (like one of many puzzles)

> Is it possible to import other extensions?

Yes, Phalcon, for example uses PSR extension (https://github.com/jbboehr/php-psr). Since version `0.14` this process started to be less painful, no need to manually include C headers of another external libraries. One thing that must be done - external extension(s) must be installed before compilation (via pecl (ex: `pecl install redis`) or OS package manager (ex: `apt install php8.0-redis`)).

**Reply by barbosa89** [reply]
Very, very great...

Thank you for taking the time to respond.

I think PHP need of Zephir in this moment.

**Reply by barbosa89** [reply]
I forget, it is possible to use yield keyword?

**Reply by Jeckerson** [reply]
Work in progress - https://github.com/zephir-lang/zephir/issues/1849
